ipa-core/src/test_fixture/circuit.rs

```diff
@@ -1,89 +1,141 @@
+    a: Replicated<F, N>,
+    b: Vec<Replicated<F, N>>,
+}
+
+impl<F: Field + FieldSimd<N>, const N: usize> Inputs<F, N> {
+    fn new(a: Replicated<F, N>, b: Vec<Replicated<F, N>>) -> Self {
+        Self { a, b }
+    }
+}
+
+/// Generates test data for the arithmetic ciruit benchmark.
 ///
 /// # Panics
-/// panics when circuits did not produce the expected value.
-pub async fn arithmetic<F, const N: usize>(width: u32, depth: u16)
+/// On functional errors, since this is a benchmark.
+#[must_use]
+pub fn arithmetic_setup<F, const N: usize>(width: u32, depth: u16) -> [Vec<Inputs<F, N>>; 3]
 where
+    F: Field + FieldSimd<N>,
+    Standard: Distribution<F>,
+{
+    let mut rng = thread_rng();
+    let mut data = array::from_fn(|_| Vec::with_capacity(width as usize / N));
+    for _ in 0..(width / u32::try_from(N).unwrap()) {
+        let [a0, a1, a2] = [F::ONE; N].share_with(&mut rng);
+        let mut b0 = Vec::with_capacity(depth as usize);
+        let mut b1 = Vec::with_capacity(depth as usize);
+        let mut b2 = Vec::with_capacity(depth as usize);
+        for _ in 0..(depth as usize) {
+            let [s0, s1, s2] = [F::ONE; N].share_with(&mut rng);
+            b0.push(s0);
+            b1.push(s1);
+            b2.push(s2);
+        }
+        data[0].push(Inputs::new(a0, b0));
+        data[1].push(Inputs::new(a1, b1));
+        data[2].push(Inputs::new(a2, b2));
+    }
+    data
+}
+
+/// Creates an arithmetic circuit with the given width and depth.
+///
+/// # Panics
+/// On functional errors, since this is a benchmark.
+pub async fn arithmetic<F, const N: usize>(
+    width: u32,
+    depth: u16,
+    active_work: usize,
+    input_data: [Vec<Inputs<F, N>>; 3],
+) where
     F: Field + FieldSimd<N> + U128Conversions,
     for<'a> Replicated<F, N>: SecureMul<SemiHonestContext<'a>>,
     [F; N]: IntoShares<Replicated<F, N>>,
     Standard: Distribution<F>,
 {
-    let world = TestWorld::default();
+    let config = TestWorldConfig {
+        gateway_config: GatewayConfig::new(active_work),
+        ..Default::default()
+    };
+    let world = TestWorld::new_with(config);
+
     // Re-use contexts for the entire execution because record identifiers are contiguous.
     let contexts = world.contexts();
 
-    let mut multiplications = Vec::new();
-    for record in 0..width {
-        let circuit_result = circuit(&contexts, RecordId::from(record), depth);
-        multiplications.push(circuit_result);
-    }
+    let [inp0, inp1, inp2] = input_data;
+
+    let Ok([fut0, fut1, fut2]): Result<[_; 3], _> = contexts
+        .into_iter()
+        .zip([inp0, inp1, inp2])
+        .map(|(ctx, col_data)| {
+            // Setting TotalRecords::Indeterminate causes OrderingSender to make data available to
+            // the channel immediately, instead of doing so only after active_work records have
+            // accumulated. This gives the best performance for vectorized operation.
```

Comment:
This is a little bit of a concern.

Reply:
We probably want to do one or more of the following:

1. Decouple OrderingSender capacity from send threshold.
2. Decouple active_work for seq_join from OrderingSender capacity.
3. Allow out-of-order writes into the OrderingSender buffer.

Setting TotalRecords::Indeterminate is a crude way of doing #&#x2060;1. I'd like to get some data from a more realistic test case before doing more than that.